src/async_fuse/memfs/cache/block.rs

```diff
@@ -0,0 +1,306 @@
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Get a mutable slice of the underlying data, copy them if there are other blocks hold the same data with `Arc`.
+    /// See also [`Arc::make_mut`](fn@std::sync::Arc::make_mut).
+    pub fn make_mut(&mut self) -> &mut [u8] {
+        Arc::make_mut(&mut self.inner).as_mut()
+    }
+
+    /// Checks if the block is dirty.
+    #[must_use]
+    pub fn dirty(&self) -> bool {
+        self.dirty
+    }
+
+    /// Sets the block to be dirty.
+    pub fn set_dirty(&mut self) {
+        self.dirty = true;
+    }
+}
+
+/// A wrapper for `IoBlock`, which is used for I/O operations
+#[derive(Clone)]
+pub struct IoBlock {
+    /// The inner `Block` that contains data
+    inner: Block,
+    /// The offset for this `Block`
+    offset: usize,
+    /// The end offset for this `Block`
+    end: usize,
+}
+
+impl std::fmt::Debug for IoBlock {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let inner_slice = make_partial_u8_slice(self.as_slice(), 0, 8);
+
+        let result = format!(
+            "IoBlock {{ inner: {inner_slice}, offset: {}, end: {}, dirty: {} }}",
+            self.offset,
+            self.end,
+            self.dirty()
+        );
+
+        write!(f, "{result}")
+    }
+}
+
+impl IoBlock {
+    /// The constructor of `IoBlock`
+    #[must_use]
+    pub fn new(inner: Block, offset: usize, end: usize) -> Self {
+        debug_assert!(offset <= end);
+        debug_assert!(
+            end <= inner.len(),
+            "The end {end} of slice is out of range of the inner block."
+        );
+        Self { inner, offset, end }
+    }
+
+    /// The inner block
+    #[must_use]
+    pub fn block(&self) -> &Block {
+        &self.inner
+    }
+
+    /// The offset of valid bytes of the inner block
+    #[must_use]
+    pub const fn offset(&self) -> usize {
+        self.offset
+    }
+
+    /// The end offset of valid bytes of the inner block
+    #[must_use]
+    pub const fn end(&self) -> usize {
+        self.end
+    }
+
+    /// Turn `IoBlock` into slice
+    pub(crate) fn as_slice(&self) -> &[u8] {
+        self.inner
+            .inner
+            .get(self.offset..self.end)
+            .unwrap_or_else(|| {
+                unreachable!(
+                    "`{}..{}` is checked not to be out of range of inner block.",
+                    self.offset, self.end,
+                )
+            })
+    }
+
+    /// Checks if the inner block is dirty.
+    #[must_use]
+    pub fn dirty(&self) -> bool {
+        self.inner.dirty
+    }
+
+    /// Sets the inner block to be dirty.
+    pub fn set_dirty(&mut self) {
+        self.inner.dirty = true;
+    }
+}
+
+#[cfg(test)]
+impl From<Block> for IoBlock {
+    fn from(block: Block) -> Self {
+        let len = block.len();
+        IoBlock::new(block, 0, len)
+    }
+}
+
+impl CouldBeAsIoVecList for IoBlock {}
+
+impl AsIoVec for IoBlock {
```

Comment:
Does  the trait `AsIoVec` make sense now?
I am not sure about its usage is outdated.

Reply:
I think it's OK and there is no need to change it for now.